tapchannel/commitment.go

```diff
@@ -0,0 +1,1129 @@
+				if err != nil {
+					err = fmt.Errorf("unable to decode "+
+						"asset htlc: %w", err)
+					return
+				}
+
+				decodedEntry := &DecodedDescriptor{
+					PaymentDescriptor: entry,
+					AssetBalances:     assetHtlc.Balances(),
+				}
+
+				local, remote = processRemoveEntry(
+					decodedEntry, local, remote,
+					isOurCommit, true, nextHeight,
+				)
+			})
+			if err != nil {
+				return 0, 0, nil, fmt.Errorf("error "+
+					"processing our remove entry: %w", err)
+			}
+		}
+	}
+	for _, entry := range original.TheirUpdates {
+		switch entry.EntryType {
+		// Skip adds for now, they will be processed below.
+		case lnwallet.Add:
+			continue
+
+		// Fee updates don't concern us at the asset level.
+		case lnwallet.FeeUpdate:
+			continue
+
+		// A settle or a timeout means we need to skip the corresponding
+		// "add" entry.
+		case lnwallet.Settle, lnwallet.Fail, lnwallet.MalformedFail:
+			skipUs[entry.ParentIndex] = struct{}{}
+
+			parentEntry, ok := localHtlcIndex[entry.ParentIndex]
+			if !ok {
+				return 0, 0, nil, fmt.Errorf("unable to find "+
+					"local htlc with index %d",
+					entry.ParentIndex)
+			}
+
+			var err error
+			parentEntry.CustomRecords.WhenSome(func(blob tlv.Blob) {
+				var assetHtlc *cmsg.Htlc
+				assetHtlc, err = cmsg.DecodeHtlc(blob)
+				if err != nil {
+					err = fmt.Errorf("unable to decode "+
+						"asset htlc: %w", err)
+					return
+				}
+
+				decodedEntry := &DecodedDescriptor{
+					PaymentDescriptor: entry,
+					AssetBalances:     assetHtlc.Balances(),
+				}
+				local, remote = processRemoveEntry(
+					decodedEntry, local, remote,
+					isOurCommit, false, nextHeight,
+				)
+			})
+			if err != nil {
+				return 0, 0, nil, fmt.Errorf("error "+
+					"processing their remove entry: %w",
+					err)
+			}
+		}
+	}
+
+	// Next we take a second pass through all the log entries, skipping any
+	// settled HTLCs, and debiting the chain state balance due to any newly
+	// added HTLCs.
+	for _, entry := range original.OurUpdates {
+		isAdd := entry.EntryType == lnwallet.Add
+
+		// Skip any entries that aren't adds or adds that were already
+		// settled or failed by a child HTLC entry we processed above.
+		if _, ok := skipUs[entry.HtlcIndex]; !isAdd || ok {
+			continue
+		}
+
+		// Again skip any entries that aren't TAP related.
+		if entry.CustomRecords.IsNone() {
+			continue
+		}
+
+		var err error
+		entry.CustomRecords.WhenSome(func(blob tlv.Blob) {
+			var assetHtlc *cmsg.Htlc
+			assetHtlc, err = cmsg.DecodeHtlc(blob)
+			if err != nil {
+				err = fmt.Errorf("unable to decode asset "+
+					"htlc: %w", err)
+				return
+			}
+
+			decodedEntry := &DecodedDescriptor{
+				PaymentDescriptor: entry,
+				AssetBalances:     assetHtlc.Balances(),
+			}
+			local, remote = processAddEntry(
+				decodedEntry, local, remote, isOurCommit, false,
+				nextHeight,
+			)
+
+			newView.OurUpdates = append(
+				newView.OurUpdates, decodedEntry,
+			)
```

Comment:
I tried this with the `processRemoveEntry` further up and it would look something like this:

```go
			var assetHtlc lfn.Option[rfqmsg.Htlc]
			err := lfn.MapOptionZ(
				parentEntry.CustomRecords,
				func(blob tlv.Blob) error {
					htlc, err := rfqmsg.DecodeHtlc(blob)
					if err != nil {
						return fmt.Errorf("unable to "+
							"decode asset htlc: %w",
							err)
					}

					assetHtlc = lfn.Some(*htlc)

					return nil
				},
			)
			if err != nil {
				return 0, 0, nil, fmt.Errorf("error "+
					"processing our remove entry: %w", err)
			}

			assetHtlc.WhenSome(func(a rfqmsg.Htlc) {
				decodedEntry := &DecodedDescriptor{
					PaymentDescriptor: entry,
					AssetBalances:     a.Balances(),
				}

				local, remote = processRemoveEntry(
					decodedEntry, local, remote,
					isOurCommit, true, nextHeight,
				)
			})
```

And in my opinion it requires more mental work to parse what's happening compared to having everything inside a single `WhenSome`.